Add Aqara H2 switches (EU+US)

lonevvolf opened the pull request:
## Proposed change
Added partial support for [Aqara Light Switch H2 US](https://www.aqara.com/us/product/light-switch-h2-us/) and [Aqara Light Switch H2 EU](https://eu.aqara.com/products/aqara-light-switch-h2). I only have the EU versions so I only added these two, but I took the proposed change from https://github.com/zigpy/zha-device-handlers/pull/4008 for the US version since that PR seems to be stuck.


## Additional information


## Checklist
- [x] The changes are tested and work correctly
- [x] `pre-commit` checks pass / the code has been formatted using Black
- [x] Tests have been added to verify that the new code works

lonevvolf committed:
Create switch_h2.py

Added first version, based on work from https://github.com/zigpy/zha-device-handlers/pull/4008

lonevvolf committed:
Merge branch 'zigpy:dev' into dev

pre-commit-ci[bot] committed:
Apply pre-commit auto fixes

ryan-adler committed:
Added test for PowerMeasurementCluster

lonevvolf committed:
Merge branch 'dev' into dev

pre-commit-ci[bot] committed:
Apply pre-commit auto fixes

lonevvolf committed:
Merge branch 'dev' into dev

EuleMitKeule commented:
Hi, I have rewritten this quirk with the V2 syntax and added the missing configuration entities/clusters that are present in the zigbee-herdsman converter.
Everything seems to be working, however I haven't yet found out how I can reorder and rename the main switch entities to `switch_left` and `switch_right`.

Maybe you could use my code and check if I missed something or if something could be improved (like renaming the main switches), and then commit it to this PR?

https://gist.github.com/EuleMitKeule/b21d32f7a0ac5788046ae6730d1e9c9f

PS: If you decide to use this, It would be nice if you could mention me as a co-author in the commit message/description :)

lonevvolf commented:
@EuleMitKeule This is great work!  Thank you so much - I will work on converting all of the devices in the quirk to the new syntax, since it looks like you've converted one so far.  Too bad there's no decent documentation on the v2 quirks.  Makes me wonder how people are expected to implement this in a consistent way.

I only have two of the 5 implemented devices there - the others are US versions or differently wired so they present different clusters.  And actually there's still one missing (EU 1 button with neutral) which isn't even in the quirk file yet.

EuleMitKeule commented:
>  Too bad there's no decent documentation on the v2 quirks

Yup, I had to look at the `zigpy`, `zha-device-handlers`, `zigbee-herdsman-converters` and `zha` repos code bases to look up everything. The most important part is the `QuirkBuilder` class in `zigpy`.  There are also some Aqara devices already implemented with quirks v2 in this repo like `remote_h1.py` I think.

> I will work on converting all of the devices

If they are identical I think there is a method `also_applies_to` or something like that in the quirk builder that allows to specify another manufacturer and model.
For the other ones, Maybe you could write a method that takes the different clusters and features as arguments and then conditionally applies the QuirkBuilder methods, so you don't have to copy/paste the whole QuirkBuilder code multiple times.

lonevvolf commented:
Right, it looks like there's no signature validation in v2?  That means we need another way to distinguish between two signatures for the same device.  Will try to look into it later on.

EuleMitKeule commented:
I didn't even know that identical device models with different signatures were a thing. Don't think I can help with that.
If you need help with anything else though please let me know, I'll have some more time at the end of the week.

Btw maybe you want to convert this PR to a draft, so it doesn't accidentally get merged?

lonevvolf commented:
Well, it would be nice if it gets merged. I expect it can be updated with more functionality later on and moved to v2.  But due to the lack of docs around v2, I don’t understand how questions like multiple signatures for the same device are supposed to be answered.

EuleMitKeule commented:
Makes sense. Another idea would be to introduce the missing features but without using the v2 syntax for now.

lonevvolf commented:
I'll try to look at that by this weekend.

einord commented:
> I'll try to look at that by this weekend.

Famous last words

lonevvolf committed:
Fix for decoupled mode

yavanosta commented:
Hi! Are there any updates on this PR? Is any help needed?

ryan-adler commented:
Any help would be greatly appreciated.  I think they would like it migrated to v2, but due to lack of documentation (and time) I have not been able to convert it.

EuleMitKeule commented:
> Hi, I have rewritten this quirk with the V2 syntax and added the missing configuration entities/clusters that are present in the zigbee-herdsman converter. Everything seems to be working, however I haven't yet found out how I can reorder and rename the main switch entities to `switch_left` and `switch_right`.
> 
> Maybe you could use my code and check if I missed something or if something could be improved (like renaming the main switches), and then commit it to this PR?
> 
> https://gist.github.com/EuleMitKeule/b21d32f7a0ac5788046ae6730d1e9c9f
> 
> PS: If you decide to use this, It would be nice if you could mention me as a co-author in the commit message/description :)

I shared a basis for the V2 code here, it's just not 100% there as I explained.

lonevvolf commented:
Looks like there's a guide on v2 quirks now (https://github.com/zigpy/zha-device-handlers/discussions/4339), I'll give it a crack and use the basis from @EuleMitKeule.

lonevvolf committed:
Moved to v2 quirks, removed US switches

Updated to v2 quirks - huge thanks to @EuleMitKeule for the initial conversion and direction.

I have removed the US version for now, as I cannot test, but will try to expand with more models support soon.

pre-commit-ci[bot] committed:
Apply pre-commit auto fixes

lonevvolf committed:
Merge branch 'dev' into dev

lonevvolf committed:
Added title string

lonevvolf committed:
Updated H2 tests for v2 quirk

pre-commit-ci[bot] committed:
Apply pre-commit auto fixes

lonevvolf committed:
Corrected capitalization

lonevvolf committed:
More capitalization

lonevvolf committed:
Fixed brand

lonevvolf committed:
Simplified test

pre-commit-ci[bot] committed:
Apply pre-commit auto fixes

lonevvolf commented:
@TheJulianJES This has been converted to v2, please advise if anything else is needed to proceed.

lonevvolf committed:
Fixed device automation triggers

pre-commit-ci[bot] committed:
Apply pre-commit auto fixes

ryan-adler committed:
Fix for tests and use of proper endpoint (12)

lonevvolf committed:
Added bind test

yavanosta commented:
Hi ZHA team?

Is there anything that can be done to expedite review of this PR?

xblitz commented:
> Hi ZHA team?
> 
> Is there anything that can be done to expedite review of this PR?

Now I understand why Z2M is what most people use, ZHA just blocks from adding new devices

marcuscramer commented:
off topic (feel free to ignore and/or direct me to the proper channel): i'm using this switch with the plain-vanilla zha settings. it randomly turns off 1-2 times a day. anyone else experiencing this?

xvzf commented:
@TheJulianJES any chance to get this moving forward? Happy to review/test/contribute/fix whatever is in the way as well

rklomp commented:
Works like a charm for me! Only thing is that with a two channel device you get two switch controls that are both called Switch and thsese get ids `switch.xxxx_switch` and `switch.xxxx_switch_2`

I have renamed them, but maybe its possible to give these more meaningful names by default. For example Switch Left and Switch Right to know which is which. Not sure if that's possible using a quirk though?

boswelja commented:
Hey! I have a 2 switch 1 channel US variant (AU, but they're identical). The device ID is different `lumi.switch.agl004`, I tried copying the configuration for `lumi.switch.agl009` and changing the ID, which got me some extra options. ~~The LED indicator switch only appears to affect the top LED, and "Flip LED indicator" doesn't seem to have any tangible effect (in relay mode, the LED indicators aren't doing anything).~~

The Z2M name is a big hint - "LED Indicator" doesn't control whether the LED indicator is currently on - it controls whether the LED indicator _can be turned on_. This isn't particularly helpful for decoupled mode, where I'd want to manually control the LED state.

I assume "Flip LED indicators" is user error, but can we add a configuration for my switch model here?

boswelja commented:
Related - z2m gives us some better option naming, if we can rename entities to provide a bit more detail that would be terrific!
There's also a "Mode select" that lets us pick from "quick_mode" or "anti_flicker_mode". That would be a good thing to expose too.
<img width="548" height="666" alt="image" src="https://github.com/user-attachments/assets/4d5762e5-f0f9-419a-b369-8508136d6950" />

janruppert commented:
Hello - thanks for the quirk-file!
I have several EU versions installed (no neutral), i.e. 2 channel 4 button (agl010) and 1 channel 2 buttons (agl009), and connect via ZHA.

I am mostly happy, however, I can't get double taps to be registered on my agl010 devices on button 1 (left top) or 2 (right top). Hold and release are also not registred...

The left relay is in decoupled mode, the right relay is in relay mode. Multi click is enabled for both sides.
Button 3 and 4 (the lower ones) don't have this issues.

Am I overlooking something or is multiclick simply not supported for the upper switches if (at least one) relay is in relay mode?

Thanks!

boswelja commented:
> Hello - thanks for the quirk-file!
> I have several EU versions installed (no neutral), i.e. 2 channel 4 button (agl010) and 1 channel 2 buttons (agl009), and connect via ZHA.
> 
> I am mostly happy, however, I can't get double taps to be registered on my agl010 devices on button 1 (left top) or 2 (right top). Hold and release are also not registred...
> 
> The left relay is in decoupled mode, the right relay is in relay mode. Multi click is enabled for both sides.
> Button 3 and 4 (the lower ones) don't have this issues.
> 
> Am I overlooking something or is multiclick simply not supported for the upper switches if (at least one) relay is in relay mode?
> 
> Thanks!

Multi click is only supported for the wireless buttons, ie not the top ones with relays (even in decoupled mode)

boswelja commented:
So I've done a bit of iteration on the US/Global variant support (I have a house full of them now)
I've mostly settled on the following additions

```py

# 2 button, 1 channel US/Global variant
(
    QuirkBuilder("Aqara", "lumi.switch.agl004")
    .adds_endpoint(1, device_type=zha.DeviceType.ON_OFF_SWITCH)
    .adds(Identify, endpoint_id=1)
    .replaces(MultistateInputCluster, endpoint_id=1)
    .replaces(MultistateInputCluster, endpoint_id=4)
    .replaces(AqaraManuSpecificCluster, endpoint_id=1)
    .replaces(AqaraManuSpecificCluster, endpoint_id=4)
    .replaces(PowerMeasurementCluster, endpoint_id=21)
    .switch(
        AqaraManuSpecificCluster.AttributeDefs.led_indicator.name,
        AqaraManuSpecificCluster.cluster_id,
        endpoint_id=1,
        translation_key="led_indicator",
        fallback_name="LED indicator",
    )
    .switch(
        AqaraManuSpecificCluster.AttributeDefs.flip_led_indicator.name,
        AqaraManuSpecificCluster.cluster_id,
        endpoint_id=1,
        translation_key="flip_led_indicator",
        fallback_name="Flip LED indicator",
    )
    .enum(
        AqaraManuSpecificCluster.AttributeDefs.power_on_mode.name,
        AqaraPowerOnMode,
        AqaraManuSpecificCluster.cluster_id,
        endpoint_id=1,
        translation_key="power_on_mode",
        fallback_name="Power on mode",
    )
    .enum(
        AqaraManuSpecificCluster.AttributeDefs.operation_mode.name,
        AqaraOperationMode,
        AqaraManuSpecificCluster.cluster_id,
        endpoint_id=1,
        translation_key="operation_mode",
        fallback_name="Operation mode",
    )
    .switch(
        AqaraManuSpecificCluster.AttributeDefs.lock_relay.name,
        AqaraManuSpecificCluster.cluster_id,
        endpoint_id=1,
        translation_key="lock_relay",
        fallback_name="Lock relay",
    )
    .switch(
        AqaraManuSpecificCluster.AttributeDefs.multi_click.name,
        AqaraManuSpecificCluster.cluster_id,
        off_value=1,
        on_value=2,
        endpoint_id=4,
        translation_key="multi_click_wireless",
        fallback_name="Multi click (wireless)",
    )
    .device_automation_triggers(
        {
            (COMMAND_SINGLE, BUTTON_1): {COMMAND: COMMAND_1_SINGLE},
            (COMMAND_HOLD, BUTTON_2): {COMMAND: COMMAND_4_HOLD},
            (COMMAND_SINGLE, BUTTON_2): {COMMAND: COMMAND_4_SINGLE},
            (COMMAND_DOUBLE, BUTTON_2): {COMMAND: COMMAND_4_DOUBLE},
            (COMMAND_RELEASE, BUTTON_2): {COMMAND: COMMAND_4_RELEASE},
        }
    )
    .add_to_registry()
)

# 2 button 2 channel US/Global variant
(
    QuirkBuilder("Aqara", "lumi.switch.agl005")
    .adds_endpoint(1, device_type=zha.DeviceType.ON_OFF_SWITCH)
    .adds_endpoint(2, device_type=zha.DeviceType.ON_OFF_SWITCH)
    .adds(Identify, endpoint_id=1)
    .replaces(MultistateInputCluster, endpoint_id=1)
    .replaces(MultistateInputCluster, endpoint_id=2)
    .replaces(AqaraManuSpecificCluster, endpoint_id=1)
    .replaces(AqaraManuSpecificCluster, endpoint_id=2)
    .replaces(PowerMeasurementCluster, endpoint_id=21)
    .switch(
        AqaraManuSpecificCluster.AttributeDefs.led_indicator.name,
        AqaraManuSpecificCluster.cluster_id,
        endpoint_id=1,
        translation_key="led_indicator",
        fallback_name="LED indicator",
    )
    .switch(
        AqaraManuSpecificCluster.AttributeDefs.flip_led_indicator.name,
        AqaraManuSpecificCluster.cluster_id,
        endpoint_id=1,
        translation_key="flip_led_indicator",
        fallback_name="Flip LED indicator",
    )
    .enum(
        AqaraManuSpecificCluster.AttributeDefs.power_on_mode.name,
        AqaraPowerOnMode,
        AqaraManuSpecificCluster.cluster_id,
        endpoint_id=1,
        translation_key="power_on_mode",
        fallback_name="Power on mode",
    )
    .enum(
        AqaraManuSpecificCluster.AttributeDefs.operation_mode.name,
        AqaraOperationMode,
        AqaraManuSpecificCluster.cluster_id,
        endpoint_id=1,
        translation_key="operation_mode_top",
        fallback_name="Operation mode (top)",
    )
    .enum(
        AqaraManuSpecificCluster.AttributeDefs.operation_mode.name,
        AqaraOperationMode,
        AqaraManuSpecificCluster.cluster_id,
        endpoint_id=2,
        translation_key="operation_mode_bottom",
        fallback_name="Operation mode (bottom)",
    )
    .switch(
        AqaraManuSpecificCluster.AttributeDefs.lock_relay.name,
        AqaraManuSpecificCluster.cluster_id,
        endpoint_id=1,
        translation_key="lock_relay_top",
        fallback_name="Lock relay (top)",
    )
    .switch(
        AqaraManuSpecificCluster.AttributeDefs.lock_relay.name,
        AqaraManuSpecificCluster.cluster_id,
        endpoint_id=2,
        translation_key="lock_relay_bottom",
        fallback_name="Lock relay (bottom)",
    )
    .device_automation_triggers(
        {
            (COMMAND_SINGLE, BUTTON_1): {COMMAND: COMMAND_1_SINGLE},
            (COMMAND_SINGLE, BUTTON_2): {COMMAND: COMMAND_4_SINGLE},
        }
    )
    .add_to_registry()
)

# 4 button, 3 channel US/Global variant
(
    QuirkBuilder("Aqara", "lumi.switch.agl006")
    .adds_endpoint(1, device_type=zha.DeviceType.ON_OFF_SWITCH)
    .adds_endpoint(2, device_type=zha.DeviceType.ON_OFF_SWITCH)
    .adds_endpoint(3, device_type=zha.DeviceType.ON_OFF_SWITCH)
    .adds(Identify, endpoint_id=1)
    .replaces(MultistateInputCluster, endpoint_id=1)
    .replaces(MultistateInputCluster, endpoint_id=2)
    .replaces(MultistateInputCluster, endpoint_id=3)
    .replaces(MultistateInputCluster, endpoint_id=4)
    .replaces(AqaraManuSpecificCluster, endpoint_id=1)
    .replaces(AqaraManuSpecificCluster, endpoint_id=2)
    .replaces(AqaraManuSpecificCluster, endpoint_id=3)
    .replaces(AqaraManuSpecificCluster, endpoint_id=4)
    .replaces(PowerMeasurementCluster, endpoint_id=21)
    .switch(
        AqaraManuSpecificCluster.AttributeDefs.led_indicator.name,
        AqaraManuSpecificCluster.cluster_id,
        endpoint_id=1,
        translation_key="led_indicator",
        fallback_name="LED indicator",
    )
    .switch(
        AqaraManuSpecificCluster.AttributeDefs.flip_led_indicator.name,
        AqaraManuSpecificCluster.cluster_id,
        endpoint_id=1,
        translation_key="flip_led_indicator",
        fallback_name="Flip LED indicator",
    )
    .enum(
        AqaraManuSpecificCluster.AttributeDefs.power_on_mode.name,
        AqaraPowerOnMode,
        AqaraManuSpecificCluster.cluster_id,
        endpoint_id=1,
        translation_key="power_on_mode",
        fallback_name="Power on mode",
    )
    .enum(
        AqaraManuSpecificCluster.AttributeDefs.operation_mode.name,
        AqaraOperationMode,
        AqaraManuSpecificCluster.cluster_id,
        endpoint_id=1,
        translation_key="operation_mode_top",
        fallback_name="Operation mode (top)",
    )
    .enum(
        AqaraManuSpecificCluster.AttributeDefs.operation_mode.name,
        AqaraOperationMode,
        AqaraManuSpecificCluster.cluster_id,
        endpoint_id=2,
        translation_key="operation_mode_middle",
        fallback_name="Operation mode (middle)",
    )
    .enum(
        AqaraManuSpecificCluster.AttributeDefs.operation_mode.name,
        AqaraOperationMode,
        AqaraManuSpecificCluster.cluster_id,
        endpoint_id=3,
        translation_key="operation_mode_bottom",
        fallback_name="Operation mode (bottom)",
    )
    .switch(
        AqaraManuSpecificCluster.AttributeDefs.lock_relay.name,
        AqaraManuSpecificCluster.cluster_id,
        endpoint_id=1,
        translation_key="lock_relay_top",
        fallback_name="Lock relay (top)",
    )
    .switch(
        AqaraManuSpecificCluster.AttributeDefs.lock_relay.name,
        AqaraManuSpecificCluster.cluster_id,
        endpoint_id=2,
        translation_key="lock_relay_middle",
        fallback_name="Lock relay (middle)",
    )
    .switch(
        AqaraManuSpecificCluster.AttributeDefs.lock_relay.name,
        AqaraManuSpecificCluster.cluster_id,
        endpoint_id=3,
        translation_key="lock_relay_bottom",
        fallback_name="Lock relay (bottom)",
    )
    .switch(
        AqaraManuSpecificCluster.AttributeDefs.multi_click.name,
        AqaraManuSpecificCluster.cluster_id,
        off_value=1,
        on_value=2,
        endpoint_id=4,
        translation_key="multi_click_wireless",
        fallback_name="Multi click (wireless)",
    )
    .device_automation_triggers(
        {
            (COMMAND_SINGLE, BUTTON_1): {COMMAND: COMMAND_1_SINGLE},
            (COMMAND_SINGLE, BUTTON_2): {COMMAND: COMMAND_4_SINGLE},
            (COMMAND_SINGLE, BUTTON_3): {COMMAND: COMMAND_4_SINGLE},
            (COMMAND_HOLD, BUTTON_4): {COMMAND: COMMAND_4_HOLD},
            (COMMAND_SINGLE, BUTTON_4): {COMMAND: COMMAND_4_SINGLE},
            (COMMAND_DOUBLE, BUTTON_4): {COMMAND: COMMAND_4_DOUBLE},
            (COMMAND_RELEASE, BUTTON_4): {COMMAND: COMMAND_4_RELEASE},
        }
    )
    .add_to_registry()
)
```

For multi-relay devices, I've labelled their controls as top/middle/bottom, and I've labelled multi-click as wireless for the button to avoid confusion.

I haven't had a chance to look into Z2M "Mode switch", I only have a single low-use light that is operating in relay mode, so it's quite low on my priority list 😅 

@lonevvolf would you be interested in appending these quirks to your existing EU switch quirks? Or would you rather I open a separate PR?

lonevvolf committed:
Added US variants

Added US variants (thanks to @boswelja)

lonevvolf committed:
Merge branch 'dev' into dev

lonevvolf committed:
Enhance H2 switch tests with multiple quirk cases

Refactor test_h2_switch and add multiple test cases for different Aqara H2 switch quirks, ensuring proper binding and attribute updates.

pre-commit-ci[bot] committed:
Apply pre-commit auto fixes

lonevvolf committed:
Refactor assertion for attribute updates

Fix assertion formatting for clarity.

pre-commit-ci[bot] committed:
Apply pre-commit auto fixes

lonevvolf committed:
Update test cases for Aqara H2 switch endpoints

Fix for new US switch tests

pre-commit-ci[bot] committed:
Apply pre-commit auto fixes

lonevvolf committed:
Refactor device initialization in H2 switch tests

pre-commit-ci[bot] committed:
Apply pre-commit auto fixes

lonevvolf committed:
Update endpoint_ids for Aqara H2 switch tests

pre-commit-ci[bot] committed:
Apply pre-commit auto fixes

yavanosta commented:
Is it possible that a small donation can expedite  review of the PR?

jsuchal commented:
I am happy to report, that this works good on Aqara H2 single EU version (no neutral installation).

Details:
- multi click works
- single click works
- Flip LED/LED indicator does no do anything

lonevvolf committed:
Merge branch 'dev' into dev

lonevvolf commented:
@TheJulianJES Can you please remove the v2 quirk candidate flag, as it has been converted to v2.  Would also appreciate if this could be merged.

boswelja commented:
> I am happy to report, that this works good on Aqara H2 single EU version (no neutral installation).
> 
> Details:
> 
>     * multi click works
> 
>     * single click works
> 
>     * Flip LED/LED indicator does no do anything

@lonevvolf you have EU switches right? Can you confirm that LED indicator controls are working as expected on your end?
They are working OK across all variants of US/AU on my side

pujux commented:
I have the `lumi.switch.agl009` (EU 2 Button 1 Channel) and it works fine with the quirk but I noticed the "Mode" (Detached/Relay) doesn't do anything. Any ideas?

buttsmcgee1 commented:
Tested with the lumi.switch.agl006 (4 button / 3 channel US/Global variant).

Buttons 2 and 3 were showing activity events correctly in Home Assistant/ZHA, but the generated device automation triggers were not firing.

Raw zha_event output showed:

- Button 2 emits: 2_single
- Button 3 emits: 3_single

However, the quirk currently maps:

python (COMMAND_SINGLE, BUTTON_2): {COMMAND: COMMAND_4_SINGLE}, (COMMAND_SINGLE, BUTTON_3): {COMMAND: COMMAND_4_SINGLE}, 

This appears to prevent the generated device automation triggers from matching correctly.

Changing the mappings to:

python (COMMAND_SINGLE, BUTTON_2): {COMMAND: COMMAND_2_SINGLE}, (COMMAND_SINGLE, BUTTON_3): {COMMAND: COMMAND_3_SINGLE}, 

resolved the issue for me and all button automation triggers now work correctly in Home Assistant.

COMMAND_3_SINGLE also appears to already exist upstream/import correctly, but wasn’t in the import list.

boswelja commented:
I've ended up doing a little tweaking on my local quirk since things were updated (which includes the changes you mentioned, good spotting!) - I'll drop my current revision here this weekend

yavanosta commented:
It's really sad that PR review takes more then a year even after it has been tested by users.

Is there anything that can be done to expedite the process?

boswelja commented:
Sorry, I forgot to drop my current quirk revision for US/Global switches

```py
"""Aqara H2 (EU) rocker switch quirks. Also see opple_switch.py for similar switches."""

from zigpy import types
from zigpy.profiles import zha
from zigpy.quirks import CustomCluster
from zigpy.quirks.v2 import QuirkBuilder
from zigpy.zcl.clusters.general import AnalogInput, Identify
from zigpy.zcl.foundation import BaseAttributeDefs, DataTypeId, ZCLAttributeDef

from zhaquirks.const import (
    BUTTON_1,
    BUTTON_2,
    BUTTON_3,
    BUTTON_4,
    CLUSTER_ID,
    COMMAND,
    COMMAND_DOUBLE,
    COMMAND_HOLD,
    COMMAND_RELEASE,
    COMMAND_SINGLE,
    ENDPOINT_ID,
    ZHA_SEND_EVENT,
)
from zhaquirks.xiaomi import XiaomiAqaraE1Cluster
from zhaquirks.xiaomi.aqara.opple_remote import (
    COMMAND_1_DOUBLE,
    COMMAND_1_HOLD,
    COMMAND_1_RELEASE,
    COMMAND_1_SINGLE,
    COMMAND_2_DOUBLE,
    COMMAND_2_HOLD,
    COMMAND_2_RELEASE,
    COMMAND_2_SINGLE,
    COMMAND_3_SINGLE,
    COMMAND_4_DOUBLE,
    COMMAND_4_HOLD,
    COMMAND_4_RELEASE,
    COMMAND_4_SINGLE,
    COMMAND_5_DOUBLE,
    COMMAND_5_HOLD,
    COMMAND_5_RELEASE,
    COMMAND_5_SINGLE,
)
from zhaquirks.xiaomi.aqara.opple_switch import MultistateInputCluster


class AqaraMultiStateInputCluster(MultistateInputCluster):
    """Aqara multi-state input cluster"""


class AqaraOperationMode(types.enum8):
    """Aqara operation mode attribute values."""

    Decoupled = 0x00
    Relay = 0x01


class AqaraPowerOnMode(types.enum8):
    """Aqara power on mode attribute values."""

    On = 0x00
    Previous = 0x01
    Off = 0x02
    Inverted = 0x03


class PowerMeasurementCluster(CustomCluster, AnalogInput):
    """Custom cluster for power measurement."""

    PRESENT_VALUE_ATTRIBUTE = 0x0055

    async def bind(self):
        """Bind cluster."""
        result = await super().bind()
        await self.configure_reporting(
            self.PRESENT_VALUE_ATTRIBUTE,
            0,
            600,
            1,
        )
        return result

    def _update_attribute(self, attrid, value):
        super()._update_attribute(attrid, value)
        if attrid == self.PRESENT_VALUE_ATTRIBUTE:
            self.listener_event(
                ZHA_SEND_EVENT,
                {
                    "type": "power_measurement",
                    "value": value,
                },
            )


class AqaraManuSpecificCluster(XiaomiAqaraE1Cluster):
    """Manufacturer-specific cluster for Lumi features."""

    class AttributeDefs(BaseAttributeDefs):
        """Attribute definitions for Aqara switch."""

        power_on_mode = ZCLAttributeDef(
            id=0x0517,
            type=AqaraPowerOnMode,
            zcl_type=DataTypeId.uint8,
            access="rw",
            is_manufacturer_specific=True,
        )

        operation_mode = ZCLAttributeDef(
            id=0x0200,
            type=AqaraOperationMode,
            zcl_type=DataTypeId.uint8,
            access="rw",
            is_manufacturer_specific=True,
        )

        led_indicator = ZCLAttributeDef(
            id=0x0203,
            type=types.Bool,
            zcl_type=DataTypeId.bool_,
            access="rw",
            is_manufacturer_specific=True,
        )

        flip_led_indicator = ZCLAttributeDef(
            id=0x00F0,
            type=types.Bool,
            zcl_type=DataTypeId.uint8,
            access="rw",
            is_manufacturer_specific=True,
        )

        lock_relay = ZCLAttributeDef(
            id=0x0285,
            type=types.Bool,
            zcl_type=DataTypeId.uint8,
            access="rw",
            is_manufacturer_specific=True,
        )

        multi_click = ZCLAttributeDef(
            id=0x0286,
            type=types.Bool,
            zcl_type=DataTypeId.uint8,
            access="rw",
            is_manufacturer_specific=True,
        )


(
    QuirkBuilder("Aqara", "lumi.switch.agl010")
    .adds_endpoint(1, device_type=zha.DeviceType.ON_OFF_SWITCH)
    .adds_endpoint(2, device_type=zha.DeviceType.ON_OFF_SWITCH)
    .adds_endpoint(4, device_type=zha.DeviceType.ON_OFF_SWITCH)
    .adds_endpoint(5, device_type=zha.DeviceType.ON_OFF_SWITCH)
    .adds_endpoint(21, device_type=zha.DeviceType.ON_OFF_SWITCH)
    .adds(Identify, endpoint_id=1)
    .replaces(AqaraMultiStateInputCluster, endpoint_id=1)
    .replaces(AqaraMultiStateInputCluster, endpoint_id=2)
    .replaces(AqaraMultiStateInputCluster, endpoint_id=4)
    .replaces(AqaraMultiStateInputCluster, endpoint_id=5)
    .replaces(AqaraManuSpecificCluster, endpoint_id=1)
    .replaces(AqaraManuSpecificCluster, endpoint_id=2)
    .replaces(AqaraManuSpecificCluster, endpoint_id=4)
    .replaces(AqaraManuSpecificCluster, endpoint_id=5)
    .replaces(PowerMeasurementCluster, endpoint_id=21)
    .switch(
        AqaraManuSpecificCluster.AttributeDefs.led_indicator.name,
        AqaraManuSpecificCluster.cluster_id,
        endpoint_id=1,
        translation_key="led_indicator",
        fallback_name="LED indicator",
    )
    .switch(
        AqaraManuSpecificCluster.AttributeDefs.flip_led_indicator.name,
        AqaraManuSpecificCluster.cluster_id,
        endpoint_id=1,
        translation_key="flip_led_indicator",
        fallback_name="Flip LED indicator",
    )
    .enum(
        AqaraManuSpecificCluster.AttributeDefs.power_on_mode.name,
        AqaraPowerOnMode,
        AqaraManuSpecificCluster.cluster_id,
        endpoint_id=1,
        translation_key="power_on_mode",
        fallback_name="Power on mode",
    )
    .enum(
        AqaraManuSpecificCluster.AttributeDefs.operation_mode.name,
        AqaraOperationMode,
        AqaraManuSpecificCluster.cluster_id,
        endpoint_id=1,
        translation_key="operation_mode_left",
        fallback_name="Operation mode left",
        unique_id_suffix="left",
    )
    .enum(
        AqaraManuSpecificCluster.AttributeDefs.operation_mode.name,
        AqaraOperationMode,
        AqaraManuSpecificCluster.cluster_id,
        endpoint_id=2,
        translation_key="operation_mode_right",
        fallback_name="Operation mode right",
        unique_id_suffix="right",
    )
    .switch(
        AqaraManuSpecificCluster.AttributeDefs.lock_relay.name,
        AqaraManuSpecificCluster.cluster_id,
        endpoint_id=1,
        translation_key="lock_relay_left",
        fallback_name="Lock relay left",
        unique_id_suffix="left",
    )
    .switch(
        AqaraManuSpecificCluster.AttributeDefs.lock_relay.name,
        AqaraManuSpecificCluster.cluster_id,
        endpoint_id=2,
        translation_key="lock_relay_right",
        fallback_name="Lock relay right",
        unique_id_suffix="right",
    )
    .switch(
        AqaraManuSpecificCluster.AttributeDefs.multi_click.name,
        AqaraManuSpecificCluster.cluster_id,
        off_value=1,
        on_value=2,
        endpoint_id=4,
        translation_key="multi_click_left",
        fallback_name="Multi click left",
        unique_id_suffix="left",
    )
    .switch(
        AqaraManuSpecificCluster.AttributeDefs.multi_click.name,
        AqaraManuSpecificCluster.cluster_id,
        off_value=1,
        on_value=2,
        endpoint_id=5,
        translation_key="multi_click_right",
        fallback_name="Multi click right",
        unique_id_suffix="right",
    )
    .device_automation_triggers(
        {
            (COMMAND_HOLD, BUTTON_1): {COMMAND: COMMAND_1_HOLD},
            (COMMAND_SINGLE, BUTTON_1): {COMMAND: COMMAND_1_SINGLE},
            (COMMAND_DOUBLE, BUTTON_1): {COMMAND: COMMAND_1_DOUBLE},
            (COMMAND_RELEASE, BUTTON_1): {COMMAND: COMMAND_1_RELEASE},
            (COMMAND_HOLD, BUTTON_2): {COMMAND: COMMAND_2_HOLD},
            (COMMAND_SINGLE, BUTTON_2): {COMMAND: COMMAND_2_SINGLE},
            (COMMAND_DOUBLE, BUTTON_2): {COMMAND: COMMAND_2_DOUBLE},
            (COMMAND_RELEASE, BUTTON_2): {COMMAND: COMMAND_2_RELEASE},
            (COMMAND_HOLD, BUTTON_3): {COMMAND: COMMAND_4_HOLD},
            (COMMAND_SINGLE, BUTTON_3): {COMMAND: COMMAND_4_SINGLE},
            (COMMAND_DOUBLE, BUTTON_3): {COMMAND: COMMAND_4_DOUBLE},
            (COMMAND_RELEASE, BUTTON_3): {COMMAND: COMMAND_4_RELEASE},
            (COMMAND_HOLD, BUTTON_4): {COMMAND: COMMAND_5_HOLD},
            (COMMAND_SINGLE, BUTTON_4): {COMMAND: COMMAND_5_SINGLE},
            (COMMAND_DOUBLE, BUTTON_4): {COMMAND: COMMAND_5_DOUBLE},
            (COMMAND_RELEASE, BUTTON_4): {COMMAND: COMMAND_5_RELEASE},
        }
    )
    .add_to_registry()
)

(
    QuirkBuilder("Aqara", "lumi.switch.agl009")
    .adds_endpoint(1, device_type=zha.DeviceType.ON_OFF_SWITCH)
    .adds_endpoint(4, device_type=zha.DeviceType.ON_OFF_SWITCH)
    .adds_endpoint(21, device_type=zha.DeviceType.ON_OFF_SWITCH)
    .adds(Identify, endpoint_id=1)
    .replaces(AqaraMultiStateInputCluster, endpoint_id=1)
    .replaces(AqaraMultiStateInputCluster, endpoint_id=4)
    .replaces(AqaraManuSpecificCluster, endpoint_id=1)
    .replaces(PowerMeasurementCluster, endpoint_id=21)
    .switch(
        AqaraManuSpecificCluster.AttributeDefs.led_indicator.name,
        AqaraManuSpecificCluster.cluster_id,
        endpoint_id=1,
        translation_key="led_indicator",
        fallback_name="LED indicator",
    )
    .switch(
        AqaraManuSpecificCluster.AttributeDefs.flip_led_indicator.name,
        AqaraManuSpecificCluster.cluster_id,
        endpoint_id=1,
        translation_key="flip_led_indicator",
        fallback_name="Flip LED indicator",
    )
    .enum(
        AqaraManuSpecificCluster.AttributeDefs.power_on_mode.name,
        AqaraPowerOnMode,
        AqaraManuSpecificCluster.cluster_id,
        endpoint_id=1,
        translation_key="power_on_mode",
        fallback_name="Power on mode",
    )
    .enum(
        AqaraManuSpecificCluster.AttributeDefs.operation_mode.name,
        AqaraOperationMode,
        AqaraManuSpecificCluster.cluster_id,
        endpoint_id=1,
        translation_key="operation_mode",
        fallback_name="Operation mode",
    )
    .switch(
        AqaraManuSpecificCluster.AttributeDefs.lock_relay.name,
        AqaraManuSpecificCluster.cluster_id,
        endpoint_id=1,
        translation_key="lock_relay",
        fallback_name="Lock relay",
    )
    .switch(
        AqaraManuSpecificCluster.AttributeDefs.multi_click.name,
        AqaraManuSpecificCluster.cluster_id,
        off_value=1,
        on_value=2,
        endpoint_id=4,
        translation_key="multi_click",
        fallback_name="Multi click",
    )
    .device_automation_triggers(
        {
            (COMMAND_HOLD, BUTTON_1): {COMMAND: COMMAND_1_HOLD},
            (COMMAND_SINGLE, BUTTON_1): {COMMAND: COMMAND_1_SINGLE},
            (COMMAND_DOUBLE, BUTTON_1): {COMMAND: COMMAND_1_DOUBLE},
            (COMMAND_RELEASE, BUTTON_1): {COMMAND: COMMAND_1_RELEASE},
            (COMMAND_HOLD, BUTTON_2): {COMMAND: COMMAND_4_HOLD},
            (COMMAND_SINGLE, BUTTON_2): {COMMAND: COMMAND_4_SINGLE},
            (COMMAND_DOUBLE, BUTTON_2): {COMMAND: COMMAND_4_DOUBLE},
            (COMMAND_RELEASE, BUTTON_2): {COMMAND: COMMAND_4_RELEASE},
        }
    )
    .add_to_registry()
)

# 2 button, 1 channel US/Global variant
(
    QuirkBuilder("Aqara", "lumi.switch.agl004")
    .adds_endpoint(1, device_type=zha.DeviceType.ON_OFF_SWITCH)
    .adds(Identify, endpoint_id=1)
    .replaces(AqaraMultiStateInputCluster, endpoint_id=1)
    .replaces(AqaraMultiStateInputCluster, endpoint_id=4)
    .replaces(AqaraManuSpecificCluster, endpoint_id=1)
    .replaces(AqaraManuSpecificCluster, endpoint_id=4)
    .replaces(PowerMeasurementCluster, endpoint_id=21)
    .switch(
        AqaraManuSpecificCluster.AttributeDefs.led_indicator.name,
        AqaraManuSpecificCluster.cluster_id,
        endpoint_id=1,
        translation_key="led_indicator",
        fallback_name="LED indicator",
    )
    .switch(
        AqaraManuSpecificCluster.AttributeDefs.flip_led_indicator.name,
        AqaraManuSpecificCluster.cluster_id,
        endpoint_id=1,
        translation_key="flip_led_indicator",
        fallback_name="Flip LED indicator",
    )
    .enum(
        AqaraManuSpecificCluster.AttributeDefs.power_on_mode.name,
        AqaraPowerOnMode,
        AqaraManuSpecificCluster.cluster_id,
        endpoint_id=1,
        translation_key="power_on_mode",
        fallback_name="Power on mode",
    )
    .enum(
        AqaraManuSpecificCluster.AttributeDefs.operation_mode.name,
        AqaraOperationMode,
        AqaraManuSpecificCluster.cluster_id,
        endpoint_id=1,
        translation_key="operation_mode",
        fallback_name="Operation mode",
    )
    .switch(
        AqaraManuSpecificCluster.AttributeDefs.lock_relay.name,
        AqaraManuSpecificCluster.cluster_id,
        endpoint_id=1,
        translation_key="lock_relay",
        fallback_name="Lock relay",
    )
    .switch(
        AqaraManuSpecificCluster.AttributeDefs.multi_click.name,
        AqaraManuSpecificCluster.cluster_id,
        off_value=1,
        on_value=2,
        endpoint_id=4,
        translation_key="multi_click",
        fallback_name="Multi click (wireless)",
    )
    .device_automation_triggers(
        {
            (COMMAND_SINGLE, BUTTON_1): {COMMAND: COMMAND_1_SINGLE},
            (COMMAND_HOLD, BUTTON_2): {COMMAND: COMMAND_4_HOLD},
            (COMMAND_SINGLE, BUTTON_2): {COMMAND: COMMAND_4_SINGLE},
            (COMMAND_DOUBLE, BUTTON_2): {COMMAND: COMMAND_4_DOUBLE},
            (COMMAND_RELEASE, BUTTON_2): {COMMAND: COMMAND_4_RELEASE},
        }
    )
    .add_to_registry()
)

# 2 button 2 channel US/Global variant
(
    QuirkBuilder("Aqara", "lumi.switch.agl005")
    .adds_endpoint(1, device_type=zha.DeviceType.ON_OFF_SWITCH)
    .adds_endpoint(2, device_type=zha.DeviceType.ON_OFF_SWITCH)
    .adds(Identify, endpoint_id=1)
    .replaces(AqaraMultiStateInputCluster, endpoint_id=1)
    .replaces(AqaraMultiStateInputCluster, endpoint_id=2)
    .replaces(AqaraManuSpecificCluster, endpoint_id=1)
    .replaces(AqaraManuSpecificCluster, endpoint_id=2)
    .replaces(PowerMeasurementCluster, endpoint_id=21)
    .switch(
        AqaraManuSpecificCluster.AttributeDefs.led_indicator.name,
        AqaraManuSpecificCluster.cluster_id,
        endpoint_id=1,
        translation_key="led_indicator",
        fallback_name="LED indicator",
    )
    .switch(
        AqaraManuSpecificCluster.AttributeDefs.flip_led_indicator.name,
        AqaraManuSpecificCluster.cluster_id,
        endpoint_id=1,
        translation_key="flip_led_indicator",
        fallback_name="Flip LED indicator",
    )
    .enum(
        AqaraManuSpecificCluster.AttributeDefs.power_on_mode.name,
        AqaraPowerOnMode,
        AqaraManuSpecificCluster.cluster_id,
        endpoint_id=1,
        translation_key="power_on_mode",
        fallback_name="Power on mode",
    )
    .enum(
        AqaraManuSpecificCluster.AttributeDefs.operation_mode.name,
        AqaraOperationMode,
        AqaraManuSpecificCluster.cluster_id,
        endpoint_id=1,
        translation_key="operation_mode_top",
        fallback_name="Operation mode (top)",
    )
    .enum(
        AqaraManuSpecificCluster.AttributeDefs.operation_mode.name,
        AqaraOperationMode,
        AqaraManuSpecificCluster.cluster_id,
        endpoint_id=2,
        translation_key="operation_mode_bottom",
        fallback_name="Operation mode (bottom)",
    )
    .switch(
        AqaraManuSpecificCluster.AttributeDefs.lock_relay.name,
        AqaraManuSpecificCluster.cluster_id,
        endpoint_id=1,
        translation_key="lock_relay_top",
        fallback_name="Lock relay (top)",
    )
    .switch(
        AqaraManuSpecificCluster.AttributeDefs.lock_relay.name,
        AqaraManuSpecificCluster.cluster_id,
        endpoint_id=2,
        translation_key="lock_relay_bottom",
        fallback_name="Lock relay (bottom)",
    )
    .device_automation_triggers(
        {
            (COMMAND_SINGLE, BUTTON_1): {COMMAND: COMMAND_1_SINGLE},
            (COMMAND_SINGLE, BUTTON_2): {COMMAND: COMMAND_2_SINGLE},
        }
    )
    .add_to_registry()
)

# 4 button, 3 channel US/Global variant
(
    QuirkBuilder("Aqara", "lumi.switch.agl006")
    .adds_endpoint(1, device_type=zha.DeviceType.ON_OFF_SWITCH)
    .adds_endpoint(2, device_type=zha.DeviceType.ON_OFF_SWITCH)
    .adds_endpoint(3, device_type=zha.DeviceType.ON_OFF_SWITCH)
    .adds(Identify, endpoint_id=1)
    .replaces(AqaraMultiStateInputCluster, endpoint_id=1)
    .replaces(AqaraMultiStateInputCluster, endpoint_id=2)
    .replaces(AqaraMultiStateInputCluster, endpoint_id=3)
    .replaces(AqaraMultiStateInputCluster, endpoint_id=4)
    .replaces(AqaraManuSpecificCluster, endpoint_id=1)
    .replaces(AqaraManuSpecificCluster, endpoint_id=2)
    .replaces(AqaraManuSpecificCluster, endpoint_id=3)
    .replaces(AqaraManuSpecificCluster, endpoint_id=4)
    .replaces(PowerMeasurementCluster, endpoint_id=21)
    .switch(
        AqaraManuSpecificCluster.AttributeDefs.led_indicator.name,
        AqaraManuSpecificCluster.cluster_id,
        endpoint_id=1,
        translation_key="led_indicator",
        fallback_name="LED indicator",
    )
    .switch(
        AqaraManuSpecificCluster.AttributeDefs.flip_led_indicator.name,
        AqaraManuSpecificCluster.cluster_id,
        endpoint_id=1,
        translation_key="flip_led_indicator",
        fallback_name="Flip LED indicator",
    )
    .enum(
        AqaraManuSpecificCluster.AttributeDefs.power_on_mode.name,
        AqaraPowerOnMode,
        AqaraManuSpecificCluster.cluster_id,
        endpoint_id=1,
        translation_key="power_on_mode",
        fallback_name="Power on mode",
    )
    .enum(
        AqaraManuSpecificCluster.AttributeDefs.operation_mode.name,
        AqaraOperationMode,
        AqaraManuSpecificCluster.cluster_id,
        endpoint_id=1,
        translation_key="operation_mode_top",
        fallback_name="Operation mode (top)",
    )
    .enum(
        AqaraManuSpecificCluster.AttributeDefs.operation_mode.name,
        AqaraOperationMode,
        AqaraManuSpecificCluster.cluster_id,
        endpoint_id=2,
        translation_key="operation_mode_middle",
        fallback_name="Operation mode (middle)",
    )
    .enum(
        AqaraManuSpecificCluster.AttributeDefs.operation_mode.name,
        AqaraOperationMode,
        AqaraManuSpecificCluster.cluster_id,
        endpoint_id=3,
        translation_key="operation_mode_bottom",
        fallback_name="Operation mode (bottom)",
    )
    .switch(
        AqaraManuSpecificCluster.AttributeDefs.lock_relay.name,
        AqaraManuSpecificCluster.cluster_id,
        endpoint_id=1,
        translation_key="lock_relay_top",
        fallback_name="Lock relay (top)",
    )
    .switch(
        AqaraManuSpecificCluster.AttributeDefs.lock_relay.name,
        AqaraManuSpecificCluster.cluster_id,
        endpoint_id=2,
        translation_key="lock_relay_middle",
        fallback_name="Lock relay (middle)",
    )
    .switch(
        AqaraManuSpecificCluster.AttributeDefs.lock_relay.name,
        AqaraManuSpecificCluster.cluster_id,
        endpoint_id=3,
        translation_key="lock_relay_bottom",
        fallback_name="Lock relay (bottom)",
    )
    .switch(
        AqaraManuSpecificCluster.AttributeDefs.multi_click.name,
        AqaraManuSpecificCluster.cluster_id,
        off_value=1,
        on_value=2,
        endpoint_id=4,
        translation_key="multi_click_wireless",
        fallback_name="Multi click (wireless)",
    )
    .device_automation_triggers(
        {
            (COMMAND_SINGLE, BUTTON_1): {COMMAND: COMMAND_1_SINGLE},
            (COMMAND_SINGLE, BUTTON_2): {COMMAND: COMMAND_2_SINGLE},
            (COMMAND_SINGLE, BUTTON_3): {COMMAND: COMMAND_3_SINGLE},
            (COMMAND_HOLD, BUTTON_4): {COMMAND: COMMAND_4_HOLD},
            (COMMAND_SINGLE, BUTTON_4): {COMMAND: COMMAND_4_SINGLE},
            (COMMAND_DOUBLE, BUTTON_4): {COMMAND: COMMAND_4_DOUBLE},
            (COMMAND_RELEASE, BUTTON_4): {COMMAND: COMMAND_4_RELEASE},
        }
    )
    .add_to_registry()
)
```

kaputcha commented:
I have the **Switch H2 (AU?)**, lumi.switch.agl006 (4-button/3-relay, no-neutral installation).

Out of the box with ZHA it exposes the three relays (endpoints 1-3) plus power/energy metering, but the fourth button (endpoint 4) isn't exposed at all, and there's no operation-mode/decouple control on any of the relays; so decoupling and the wireless button are only usable via the Aqara app.

I can see the **agl006** variant is already written into this PR, which looks like it addresses exactly those gaps. I've got the device on the bench (and an Aqara M100 to cross-check behaviour against the app if that's necessaru), so if it's helpful I'm happy to load the quirk and test the **agl006** specifically and report back.

This will be the first time I'm assisting with something like this. Is there any particular behaviour you need me to check?

boswelja commented:
I've also got a 4 button 3 relay, and am based in AU! Our switches are identical to the US variants. If you want to try it out, you'll want to verify as much of the exposed components as you can, since this quirk essentially replaces all of them with custom handlers.

kaputcha commented:
Alright. So some quick tests seem to show it's operating pretty much as expected.

### Test setup

| | |
|---|---|
| **Device** | Aqara Light Switch H2 (AU), `lumi.switch.agl006` |
| **Variant** | 4-button / 3-relay, no-neutral wiring |
| **Firmware** | `0x0000171b` |
| **Environment** | HA 2026.7.0, zha-quirks 2.1.0 |
| **Quirk** | Loaded from this PR branch |

### Results

**Relays (endpoints 1–3)** - ✅
On/off control working.
❓Power/energy metering is displayed, but no values are reported (could be my specific use case though)❓

**Decouple/operation mode** - ✅
`operation_mode` selects appear for all three relays (top/middle/bottom) and decoupling works correctly: with a relay decoupled, the physical button no longer switches its load, and the load stays  controllable via HA. Per-relay `lock_relay` switches are also present and working.

**Wireless button (endpoint 4)** - ✅
Fires clean, distinct `zha_event`s for every press type:

| Command | `press_type` | `value` |
|---|---|---|
| `4_single` | single | 1 |
| `4_double` | double | 2 |
| `4_hold` | hold | 0 |
| `4_release` | release | 255 |

All on `endpoint_id: 4`, `cluster_id: 0x0012`. Hold + release fire as a pair, so press-and-hold automations are viable.

**Decoupled relay buttons as triggers (endpoints 1–3)** - ✅ (single press only)
With a relay decoupled, its physical button fires a `zha_event` that can be used as an automation trigger while the relay stays unaffected, e.g.:
```
{button: 3, press_type: single, attr_id: 85, value: 1}
```
**Only single press is emitted** on the relay-backed buttons - no double or hold, even when decoupled. This matches the earlier discussion in this thread. (Multi-click is available only on the dedicated wireless button, endpoint 4.)

**"Multi click (wireless)" config toggle** - ✅
Confirmed this entity controls multi-click detection on the wireless button (endpoint 4): with it disabled, only single-press `zha_event`s fire; re-enabling restores double/hold/release.

**ZHA reload persistence** - ✅
All entities persist after an integration reload - no dropout, no re-pair required. Only tested once though.

### Noted behaviour

**Wireless button has no device automation triggers**
The endpoint-4 button works via raw `zha_event`, but has **no `device_automation_triggers` registered**. When building an automation via **Device**, the trigger dropdown only lists the three relays and the config entities - there are no "Button 4 press" options, so the wireless button is currently only usable by hand-writing an event trigger.

If registering device triggers for the wireless button is in scope, it would make it much more accessible for UI-based automations.

**Cosmetic: button label mismatch**
The wireless button surfaces in the UI labelled "Button 2", though it's physically the 4th button (endpoint 4). The event payload is correct - it's only the friendly label that's off.

**Cosmetic: button ordering labels**
Suggest buttons be renamed to "One, Two and Three". Top, Middle, Bottom does not map well for the Horizontal version of these switch plates.

---

Happy to test anything else.

boswelja commented:
> Power/energy metering is displayed, but no values are reported

Yup, power metering is not supported on these (or any) switches without a neutral wire. If that's something you want, you'll need to run a neutral.
Less well documented, there's a minimum load requirement of 5 watts for no-neutral switches. If your connected load drops below that the switches will go offline

> Wireless button has no device automation triggers

Strange, I've got them on my end and haven't had to write anything by hand (well, except for the quirk)
<img width="1564" height="284" alt="image" src="https://github.com/user-attachments/assets/07e3041f-45bb-410a-87fe-21d6cf00d0af" />

> Cosmetic: button label mismatch

I don't have any label mismatches - in fact, I have no cosmetic labels! I'm also not sure where these would come from, since they don't exist in the quirk.

> Suggest buttons be renamed to "One, Two and Three". Top, Middle, Bottom does not map well for the Horizontal version of these switch plates.

Does the horizontal version have the same device ID (`lumi.switch.agl006`)? I would've guessed they'd have a variant for those (and thus would need their own quirk)

kaputcha commented:
> Yup, power metering is not supported on these (or any) switches without a neutral wire. If that's something you want, you'll need to run a neutral.
Less well documented, there's a minimum load requirement of 5 watts for no-neutral switches. If your connected load drops below that the switches will go offline

Absolutely makes sense, didn't even consider that as the likely reason.

> Strange, I've got them on my end and haven't had to write anything by hand (well, except for the quirk)

Okay, I've just double checked and realised I was searching by target and not type. When browsing actions by type/device "button_4" options appear.

> I don't have any label mismatches - in fact, I have no cosmetic labels! I'm also not sure where these would come from, since they don't exist in the quirk.

Disregard this one. As I said, this is the first time I've ever really helping with anything remotely related to Home Assistant, let alone on Github. When I was testing earlier on, I was drafting this as I went. I could have sworn I saw the wireless button listed as "button_2". I assumed it was a holdover from the 2-gang/single Relay version of the switch. Can't replicate that anymore.

> Does the horizontal version have the same device ID (lumi.switch.agl006)? I would've guessed they'd have a variant for those (and thus would need their own quirk)

So in my ZHA diagnostics, the switch is labeled as `lumi.switch.agl006`. Is ZHA diagnostics the correct place to look for these types of IDs?

**Diagnostics for clarity - prior to applying the quirk**

"version": 2,
 "ieee": "**REDACTED**",
    "nwk": "0x220A",
    "manufacturer": "Aqara",
    "model": "lumi.switch.agl006",
    "friendly_manufacturer": "Aqara",
    "friendly_model": "lumi.switch.agl006",
    "name": "Aqara lumi.switch.agl006",
    "quirk_applied": false,
    "quirk_class": "zigpy.device.Device",
    "exposes_features": [],
    "manufacturer_code": 4447,
    "power_source": "Battery or Unknown",
    "lqi": 54,
    "rssi": null,
    "last_seen": "2026-07-04T08:34:41.225368+00:00",
    "available": true,
    "device_type": "EndDevice",
    "active_coordinator": false,
    "node_descriptor": {
      "logical_type": "EndDevice",
      "complex_descriptor_available": false,
      "user_descriptor_available": false,
      "reserved": 0,
      "aps_flags": 0,
      "frequency_band": 8,
      "mac_capability_flags": 128,
      "manufacturer_code": 4447,
      "maximum_buffer_size": 82,
      "maximum_incoming_transfer_size": 82,
      "server_mask": 11264,
      "maximum_outgoing_transfer_size": 82,
      "descriptor_capability_field": 0
